### PR TITLE
Fix https://github.com/JuliaEarth/GeoStats.jl/issues/185

### PR DESCRIPTION
**Pull request by Datseris**
Closes https://github.com/JuliaEarth/GeoStats.jl/issues/185

### CODE REVIEW

File: src/partition.jl [hunk 1]
```diff
@@ -24,10 +24,12 @@ function EmpiricalVariogram(partition::Partition, var₁::Symbol, var₂::Symbol
 end
 
 """
-    DirectionalVariogram(direction, data, var₁, var₂=var₁; dtol=1e-6, [parameters])
+    DirectionalVariogram(direction, data, var₁::Symbol, var₂=var₁; dtol=1e-6, [parameters])
```
Comment (juliohm) [inline src/partition.jl:27]: I am wondering how we could avoid adding type annotations everywhere and still be clear. For someone using the stack, these arguments should be clear, but I agree with you that first time users may be lost without type annotations.

Comment (Datseris) [reply src/partition.jl:27]: Actually, this isn't needed. I reference below that `var`s are symbols!

**Comment by juliohm** [comment]
> Given that the Neighbors ball search only works with Minkowski metrics, shouldn't this directly error instead of giving a warning?

I think that is a good idea. We should probably just throw an error instead of warning. Users will then choose the algorithm explicitly in a subsequent call. Feel free to update the PR with the suggested change.

I am still confused about how we can improve the docstring without polluting it with types. Thinking twice about it, perhaps we should assume that anyone using the project would have written the docs and will implicitly know what data and var is from context. I have geoscientists with minimum programming experience in mind, and they get scared when they see type annotations. Also, I think docstrings should be as clean as possible and talk to the scientists as opposed to the programmers. What do you feel? Any suggestion that could address both types of audience? Maybe a glossary in the docs with terminology and conventions?

**Comment by juliohm** [comment]
Can I suggest that we write a glossary in the official docs instead and point users there? I think all docstrings across the entire project use the terminology consistently and we just need a central place with the definitions instead of repeating everywhere.

**Comment by juliohm** [comment]
We could then open a separate PR to fix the warn vs. error issue.

**Comment by Datseris** [comment]
Personally, I believe that documentation strings should be as precise as possible. You could go the "Glossary" route, that's okay, but then you really need to use special names. `data` is an extremely generic name that GeoStats.jl does not have an exclusivo on. It does have an exclusivo on its types. Saying `data::GeoRef` (don't know the type here) makes it clear what `data` is and leaves precisely zero margin for ambiguity. Equivalently, if you don't like the type approach, you can state in the docstring that `data` is the output of `georef`, as we do now in this PR.

It would be hard to establish in the mind of the user that such a generic name like `data` or `var` as an entity that should be understood only within GeoStats.jl content. Types on the other hand do make this possible.

I realized this when I added the explanation: "`var`s are symbols ....". This is imprecise and can lead to ambiguity. The precise version is "`var`s are `Symbol`s...". They are not "symbols", which could mean whatever every person has in their head when they think of symbols. They are quite precisely a `Symbol` instance of the Julia language. If we were expecting this package to be used by people how are truly clueless about programming, writing "symbol" or "`Symbol`" would have no difference for them. They still wouldn't know what to do as "symbol" could mean anything for them. For people who do know the very basics, `Symbol` already makes a difference of clarity.

> Also, I think docstrings should be as clean as possible and talk to the scientists as opposed to the programmers

I think types are straightforward, and would not agree that scientists would not understand them. At least not true for the scientists I know. Types are a universal concept in programming, not really bound to Julia either.

**Comment by juliohm** [comment]
I mean we are very privileged scientists with very high-education :) I am talking about young students who know very little about programming, students coming from geology and other fields who have basically zero programming experience in the vast majority. I will try to write a big section of the docs explaining the terminology and concepts across the stack and then we can come back later to figure out what to do with the docstrings. Perhaps just adding a link to the glossary would be enough as a footenote everywhere. This could even be automated with a script to add a footnote like:

```julia
"""

For more context on arguments, please read the [glossary](link/to/official-docs).
"""
```

The glossary can be as precise as needed with a lot of text to remove ambiguity. I think docstring should be short and clean. They have a different purpose for those who just want to remember the arguments and options available.

**Comment by Datseris** [comment]
> For more context on arguments, please read the [glossary](link/to/official-docs).

> The glossary can be as precise as needed with a lot of text to remove ambiguity. I think docstring should be short and clean. They have a different purpose for those who just want to remember the arguments and options available.

Yes this is a very nice approach.

**Pull request closed without merging**